torch/testing/_internal/common_methods_invocations.py

```diff
@@ -3158,11 +3158,11 @@ def method_tests():
         ('cumprod', (S, S, S), (0,)),
         ('cumprod', (S, S, S), (1,), 'dim1', (), [0]),
         ('cumprod', (), (0,), 'scalar'),
-        ('cumprod', (torch.tensor(0., requires_grad=True)), (0,), 'scalar_zeros'),
-        ('cumprod', prod_zeros(S, [0, 1]), (1,), 'zeros_dim2', (), [0]),
-        ('cumprod', prod_zeros(S, [0, 2]), (1,), 'zeros_dim1', (), [0]),
-        ('cumprod', prod_zeros(S, [1, 2]), (1,), 'zeros_dim0', (), [0]),
-        ('cumprod', prod_zeros(S, [1, 2]), (1,), 'zeros_dim0_cast', (), [0], (), ident, {'dtype': torch.float64}),
+        ('cumprod', lambda dtype, device: (torch.tensor(0., requires_grad=True, dtype=dtype, device=device)), (0,), 'scalar_zeros'),
+        ('cumprod', lambda dtype, device: prod_zeros(S, [0, 1], dtype), (1,), 'zeros_dim2', (), [0]),
+        ('cumprod', lambda dtype, device: prod_zeros(S, [0, 2], dtype), (1,), 'zeros_dim1', (), [0]),
+        ('cumprod', lambda dtype, device: prod_zeros(S, [1, 2], dtype), (1,), 'zeros_dim0', (), [0]),
+        ('cumprod', lambda dtype, device: prod_zeros(S, [1, 2]).to(dtype), (1,), 'zeros_dim0_cast', (), [0], (), ident, {'dtype': torch.double}),
```

Comment:
@anjali411 , @mruberry , this cast test is causing issues as it does not work on master anymore. I am not sure it is reproducible with OpInfo, or am I wrong?

Reply:
@nikitaved I think you need to update the formula for `cumprod` to make it work for complex tensors.

Reply:
@Lezcano is working to add complex support for `torch.cumprod` so I think we can exclude `cumprod` related changes from this PR.

Reply:
There is no complex prod backward without `cumprod` complex support. Let' see who's PR gets landed first and then one of us could merge?

Reply:
@anjali411 , the formula for complex `cumprod_backward` is correct, the casting test is failing because it complains that `cumprod` receives and input with `self.dtype != dtype`.

Reply:
this issue is similar to https://github.com/pytorch/pytorch/pull/53240#discussion_r588326560 let's just leave that method test for now and port the rest

Reply:
I will write the `OpInfo`  tests for `cumprod`  and will add the complex support as well in my PR. I will write the benchmarks and push it later today if I don't encounter any major error when porting the tests to `OpInfo`.

As @nikita said, I'd also be happy to merge nikitaved's PR if his is merged first, as the overlapping is minimal.

Reply:
I could write tests for `prod` only, so this way we introduce no conflicts whatsoever.

Reply:
For what is worth, complex support for `cumprod` unsurprisingly depends on `prod` (among other operations). I guess this PR will have to go in first, and then I'll we can do complex `cumprod`.

Reply:
Yes, but we can decide who writes the tests for `cumprod`. I could enable its complex support, yet leave the tests intact.